docs/compiler.rst

```diff
@@ -0,0 +1,273 @@
+2. add grammar to ``scenic.gram``
+3. write parser tests
+4. add visitor to ``compiler.py``
+5. write compiler tests
+
+The rest of this section will demonstrate how we can add the ``implies``
+operator using the new parser architecture.
+
+Step 1: AST Nodes
+~~~~~~~~~~~~~~~~~
+
+First, we define AST nodes that represent the syntax. Since the
+``implies`` operator is a binary operator, the AST node will have two
+fields for each operand.
+
+.. code:: python
+
+   class ImpliesOp(AST): # 1
+       __match_args__ = ("hypothesis", "conclusion") # 2
+
+   		# 3
+       def __init__(
+           self, hypothesis: ast.AST, conclusion: ast.AST, *args: any, **kwargs: any
+       ) -> None:
+           super().__init__(*args, **kwargs)
+           self.hypothesis = hypothesis
+           self.conclusion = conclusion
+   				# 4
+           self._fields = ["hypothesis", "conclusion"]
+
+1. ``AST`` (not ``ast.AST`` ) is a Scenic base class that all Scenic AST
+   nodes extends.
+2. ``__match_args__`` is a syntax for using `structural pattern
+   matching <https://peps.python.org/pep-0636/#matching-positional-attributes>`__
+   on argument positions. This is to make it easier to write parser
+   tests.
+3. The initializer takes two operands (hypothesis and conclusion). Note
+   that their types are ``ast.AST`` , which is a base for all AST nodes,
+   including both Scenic AST nodes and Python AST nodes. ``*args`` and
+   ``**kwargs`` should be passed to the base class’ initializer to store
+   line number, offset, etc.
+4. ``_fields`` is a special field that specifies the child nodes. This is used by
+   the library functions such as ``generic_visit`` to traverse the
+   syntax tree.
+
+Step 2: Add Grammar
+~~~~~~~~~~~~~~~~~~~
+
+.. note::
+    
+    The grammar described here is simplified for the sake of brevity and is not the actual grammar used in the Scenic language.
+
+The next step is to add a grammar to the ``scenic.gram`` file.
+
+::
+
+   scenic_implication (memo):
+       | invalid_scenic_implication
+       | a=disjunction 'implies' b=disjunction { s.ImpliesOp(a, b, LOCATIONS) }
+       | disjunction
+
+In the grammar, the precedence and associativity is given by using
+separate rules for each precedence. Since implication should bind less tightly than ``or``,
+it takes two ``disjunction``\ s for its operands.
+
+All Scenic grammar rules should be prefixed with ``scenic_`` so we can
+easily distinguish rules added to the original Python grammar.
+
+The production rule is defined inside the brackets. ``s`` refers to the
+Scenic AST module and we can write the plain Python to construct the AST
+node. ``LOCATIONS`` will be replaced with a set of named arguments to
+express source code locations.
+
+The ``implies`` operator is unique in that it takes exactly two
+operands. Since the associativity of ``implies`` operator is not clear,
+developers must specify the associativity by explicitly wrapping one of
+the operand with parenthesis.
```

Comment:
... the operand[s] with ...

Reply:
fixed 4cf64494a4ef4340c82445ac97db27fe9d1dd05b